src/controllers/dataTierApplicationWebviewController.ts

```diff
@@ -0,0 +1,992 @@
+                    displayName,
+                    server: profile.server,
+                    database: profile.database,
+                    authenticationType: this.getAuthenticationTypeString(
+                        profile.authenticationType,
+                    ),
+                    userName: profile.user,
+                    isConnected,
+                    profileId,
+                });
+            }
+
+            const existingProfileIds = new Set(connections.map((conn) => conn.profileId));
+
+            // Include active connections that may not appear in the recent list
+            for (const activeConnection of Object.values(activeConnections)) {
+                const profile = activeConnection.credentials as IConnectionProfile;
+                const profileId = profile.id || `${profile.server}_${profile.database || ""}`;
+
+                if (existingProfileIds.has(profileId)) {
+                    continue;
+                }
+
+                // Only include if actually connected (not in connecting state or errored)
+                const ownerUri = this.connectionManager.getUriForConnection(profile);
+                if (!ownerUri || !this.connectionManager.isConnected(ownerUri)) {
+                    continue;
+                }
+
+                const displayName = this.buildConnectionDisplayName(profile);
+
+                connections.push({
+                    displayName,
+                    server: profile.server,
+                    database: profile.database,
+                    authenticationType: this.getAuthenticationTypeString(
+                        profile.authenticationType,
+                    ),
+                    userName: profile.user,
+                    isConnected: true,
+                    profileId,
+                });
+                existingProfileIds.add(profileId);
+            }
+
+            return { connections };
+        } catch (error) {
+            this.logger.error(`Failed to list connections: ${error}`);
+            return { connections: [] };
+        }
+    }
+
+    /**
+     * Initializes connection based on initial state from Object Explorer or previous session
+     * Handles auto-matching and auto-connecting to provide seamless user experience
+     */
+    private async initializeConnection(params: {
+        initialServerName?: string;
+        initialDatabaseName?: string;
+        initialOwnerUri?: string;
+        initialProfileId?: string;
+    }): Promise<{
+        connections: ConnectionProfile[];
+        selectedConnection?: ConnectionProfile;
+        ownerUri?: string;
+        autoConnected: boolean;
+        errorMessage?: string;
+    }> {
+        try {
+            // Get all connections (recent + active)
+            const { connections } = await this.listConnections();
+
+            // Helper to find matching connection
+            const findMatchingConnection = (): ConnectionProfile | undefined => {
```

Comment:
Check out `ConnectionStore.findMatchingProfile()`; that should already do what you need